### PR TITLE
Interop stress tests

### PR DESCRIPTION
**Pull request by sreecha**

### CODE REVIEW

File: test/cpp/interop/interop_client.cc [hunk 1]
```diff
@@ -83,7 +83,7 @@ CompressionType GetInteropCompressionTypeFromCompressionAlgorithm(
 }  // namespace
 
 InteropClient::InteropClient(std::shared_ptr<Channel> channel)
-    : channel_(channel) {}
+    : channel_(channel), stub_(TestService::NewStub(channel)) {}
```
Comment (dgquintas) [inline test/cpp/interop/interop_client.cc]: This is cleaner, but are we sure it doesn't change the semantics? Prior, the stub was re-created per testcase. This way, the same stub is reused.

File: test/cpp/interop/stress_interop_client.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    {EMPTY_STREAM, "empty_stream"}};
+
+class TestSelector {
+ public:
+  TestSelector(const vector<pair<TestCaseType, int>>& tests);
+  TestCaseType GetNextTest() const;
+
+ private:
+  const vector<pair<TestCaseType, int>> tests_;
+  int total_weight_;
```
Comment (dgquintas) [inline test/cpp/interop/stress_interop_client.h]: what's its role? Probably some docs on GetNextTest() would also be nice, its possible behaviors, etc.

File: test/cpp/interop/stress_interop_client.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+class StressTestInteropClient {
+ public:
+  StressTestInteropClient(int test_id, const string& server_address,
```
Comment (dgquintas) [inline test/cpp/interop/stress_interop_client.h]: docs pls :)

File: test/cpp/interop/stress_test.cc [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+using std::thread;
+using std::vector;
+
+using grpc::testing::kTestCaseList;
+using grpc::testing::StressTestInteropClient;
+using grpc::testing::TestCaseType;
+using grpc::testing::TestSelector;
+using grpc::testing::UNKNOWN_TEST;
+
+TestCaseType get_test_type_from_name(const string& test_name) {
```
Comment (dgquintas) [inline test/cpp/interop/stress_test.cc]: Use camelcase as per the style guide: https://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Function_Names

File: test/cpp/interop/stress_test.cc [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  for (auto it = tokens.begin(); it != tokens.end(); it++) {
+    // Token is in the form <test_name>:<test_weight>
+    size_t colon_pos = it->find(':');
+    if (colon_pos == string::npos) {
+      gpr_log(GPR_ERROR, "Error in parsing test case string: %s", it->c_str());
+      is_success = false;
+      break;
+    }
+
+    string test_name = it->substr(0, colon_pos);
```
Comment (dgquintas) [inline test/cpp/interop/stress_test.cc]: nit: const

File: test/cpp/interop/stress_test.cc [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  }
+
+  LogParameterInfo(server_addresses, tests);
+
+  TestSelector test_selector(tests);
+
+  vector<thread> test_threads;
+  int thread_idx = 0;
+  for (auto it = server_addresses.begin(); it != server_addresses.end(); it++) {
+    StressTestInteropClient* client = new StressTestInteropClient(
```
Comment (dgquintas) [inline test/cpp/interop/stress_test.cc]: prefer using std::unique_ptr

File: test/cpp/interop/stress_interop_client.h (hunk 2 above)
Comment (sreecha) [inline test/cpp/interop/stress_interop_client.h]: Renamed the class to WeightedRandomTestSelector and added a comment on what it does.

File: test/cpp/interop/stress_interop_client.h (hunk 3 above)
Comment (sreecha) [inline test/cpp/interop/stress_interop_client.h]: Done.

File: test/cpp/interop/stress_test.cc (hunk 5 above)
Comment (sreecha) [inline test/cpp/interop/stress_test.cc]: This is the output parameter that holds the parsed test case values.

File: test/cpp/interop/stress_test.cc (hunk 6 above)
Comment (sreecha) [inline test/cpp/interop/stress_test.cc]: the std::thread() constructor expects a function pointer. Which is why I passed it as 
```thread(&StressTestInteropClient::MainLoop, client)``` . I don't know to do this if client is of type unique_ptr... (i can't do client.get() obviously)

File: test/cpp/interop/stress_test.cc (hunk 4 above)
Comment (sreecha) [inline test/cpp/interop/stress_test.cc]: Done

File: test/cpp/interop/interop_client.cc (hunk 1 above)
Comment (sreecha) [inline test/cpp/interop/interop_client.cc]: Thanks for calling this out!. I spoke to Yang and looks like this is intentional - to keep the test cases clean. For interop tests this makes sense but for the stress test use case, it is not a good idea to create a stub for every call :). I am going to make this configurable.

Comment (sreecha) [inline test/cpp/interop/interop_client.cc]: I have fixed this by adding a new option to InteropClient constructor.

**Comment by sreecha** [comment]
@dgquintas - I have made the changes and pushed the latest changes. Please review them at your convenience.

File: test/cpp/interop/interop_client.cc [hunk 7]
```diff
@@ -82,8 +82,46 @@ CompressionType GetInteropCompressionTypeFromCompressionAlgorithm(
...
+  return stub_.get();
+}
+
+void InteropClient::ServiceStub::Reset(std::shared_ptr<Channel> channel) {
+  channel_ = channel;
+
+  // Update stub_ as well. Note: If new_stub_every_call_ is true, we can set
+  // stub_ to nullptr since the next call to Get() will create a new stub
+  if (new_stub_every_call_) {
+    stub_.reset(nullptr);
```
Comment (dgquintas) [inline test/cpp/interop/interop_client.cc]: you don't need to pass any argument to reset()

Comment (dgquintas) [inline test/cpp/interop/interop_client.cc]: You don't need to pass any argument to reset()

**Comment by dgquintas** [comment]
LGTM. It only needs a merge with HEAD and a jenkins green

File: test/cpp/interop/interop_client.cc (hunk 7 above)
Comment (sreecha) [inline test/cpp/interop/interop_client.cc]: Cool. Made the change.

**Comment by sreecha** [comment]
Merged with HEAD and pushed the new changes.

**Comment by dgquintas** [comment]
There are some issues, see Jenkins for details.

**Comment by sreecha** [comment]
btw, The interop tests failures look unrelated to my change.

**Pull request merged**